### PR TITLE
Implement backend for tags

### PR DESCRIPTION
**Pull request by maoivy**
Adds ```tags``` field to ```Project``` and ```Update```.

Tags are filtered, trimmed for whitespace, and set to all lowercase, with duplicates dropped, in ```router.ts```.
Removing a tag from a project will remove that tag from all updates with it.
Updates can only be given tags that are in the project's tag space.

### CODE REVIEW

File: server/update/middleware.ts [hunk 1]
```diff
@@ -228,6 +229,33 @@ import UpdateCollection from '../update/collection';
...
+    const project = await ProjectCollection.findOne(req.body.projectId);
+    const projectTags = project.tags;
+
+    for (const tag of tags) {
+      if (typeof tag !== 'string') {
+        res.status(401).json({
+          error: 'Tags must be a list of strings.'
+        });
+        return;
+      } else if (!projectTags.includes(tag)) {
```
Comment (opheeelia) [inline server/update/middleware.ts:250]: same suggestion here as prev comment

File: server/update/middleware.ts [hunk 2]
```diff
@@ -311,6 +340,34 @@ import UpdateCollection from '../update/collection';
...
+    const project = await ProjectCollection.findOne(update.projectId);
+    const projectTags = project.tags;
+
+    for (const tag of tags) {
+      if (typeof tag !== 'string') {
+        res.status(401).json({
+          error: 'Tags must be a list of strings.'
+        });
+        return;
+      } else if (!projectTags.includes(tag)) {
```
Comment (opheeelia) [inline server/update/middleware.ts:362]: same here too

**Review by opheeelia — CHANGES_REQUESTED** [review]
Can we also include a findUpdateByTag function in Update's collection.ts so that the frontend can filter by a single tag?

File: server/project/middleware.ts [hunk 3]
```diff
@@ -103,6 +105,25 @@ const isValidProjectFields = async (req: Request, res: Response, next: NextFunct
...
+  if (tags !== undefined) {
+    if (!Array.isArray(tags)) {
+      res.status(401).json({
+        error: 'Tags must be a list.'
+      });
+      return;
+    }
+
+    for (const elt of tags) {
+      if (typeof elt !== 'string') {
```
[the remark this answers is not in the document]
Comment (maoivy) [reply server/project/middleware.ts:118]: good point! I changed the project middleware to make sure tags are between 1 and 50 chars after trimming.

File: server/update/middleware.ts (hunk 1 above)
Comment (maoivy) [reply server/update/middleware.ts]: I decided to remove this completely and only do the projectTags thing—it's got to be a valid tag if it's in projectTags, so it should be enough just to check that it's in there!

**Review by opheeelia — APPROVED** [review]

**Pull request merged**